Time stretching 4 of 6 repair copy paste for stretched tracks

saintmatthieu opened the pull request:
Resolves: #4850 
Resolves: #5273

Copying a clip of a given stretch ratio onto a clip of another stretch ratio cannot work as it has been so far. Instead of rendering the time stretching, we create new clips.

Note that this PR also changes the behavior if clips have equal ratios, since this was an ask from Product (@Tantacrul @LWinterberg ) and a low-hanging fruit for this PR.

Also note that this PR removes the feature whereby clicking the junction of two clips merges them, as a provisional step towards https://github.com/audacity/audacity/issues/2330.

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

## QA

Please try a myriad of copy, paste and join operations, including trimmed clips, stretched clips, ...
Copy-paste operations should be tried with the "Editing a clip can move other clips" preference both on and off.
- [x] Cross-project copy-paste with difference of project tempo; observe time selection after paste, and stretching of the insertion which plays back at the new tempo
- [x] ~~Pasting a smart clip at the exact end of another clip preserves right-hand hidden data, loses (doesn't paste) left-hand~~ (The behaviour should be kept as in current release, that is, trimmed data should remain recoverable by untrimming).
**Peter 05Oct23 testing with latest 4 of 6 branch:**  The entire smart clip appears to be pasted. I had  30 second chirp and smart clipped it  by 10 seconds each side .  Then pasted it adjacent to  the right end of a 30 second tone.  in situ I could re-expand the chirp on the right.  I could expand it on the left in situ, but if I moved it rightwards away from the  tone I could re-expand it on the left too.  I absolutely think this is correct/desired behavior, the entire smart clip should be retained,
- [x] ~~Vice-versa, when pasting at the exact start of another clip~~
**Peter 05Oct23 testing with latest 4 of 6 branch:**  similarly so
- [x] Join stretched audio clips with equal stretch ratios doesn't render stretching
**Peter 05Oct23 testing with latest 4 of 6 branch:**  it still does render to 100%
- [x] Join audio clips with different stretch ratios shows progress bar
 **Peter 05Oct23 testing with latest 4 of 6 branch:**  Progress bar is properly shown
- [x] `Edit > Labeled Audio > Join` on project with several tracks of varying stretch ratios. (#5246)
 **Peter 05Oct23 testing with latest 4 of 6 branch:**  Can't test as labelled audio commands are FUBAR
- [x] Likewise with another label-based editing command of your choice, such as `Labeled Audio > Delete`.
 **Peter 05Oct23 testing with latest 4 of 6 branch:**  Can't test as labelled audio commands are FUBAR
- [x] Sync-lock when a wave track with stretch lengthens, inserting silence into a clip
- [x] Generating audio over a stretched clip yields expected result, be it silence, an Audacity or a Nyquist tone.
**Peter 05Oct23 testing with latest 4 of 6 branch:**  This looks to work properly in the branch - the generated audio is 100%, but the audio clips either side retain their stretch % factor (and are not rendered).  Expected and desired behavior I think.
- [x] Envelope still working, also when drawing. Please make sure the end samples are correctly scaled and editable.
- [x] Fixes https://github.com/audacity/audacity/issues/4984 and https://github.com/audacity/audacity/issues/5145 (please close if it does)
**Peter 05Oct23 testing with latest 4 of 6 branch:**  note the cliplet/"invisible" clip syndrome.
- [x] Any chance it fixes https://github.com/audacity/audacity/issues/4297 ?
**Peter 05Oct23 testing with latest 4 of 6 branch:**    Looks to be fixed in the 4 of 6 branch
- [x] Paste a clip A onto a portion of a clip B. Now move A out of the way. The data in B that was under A is recoverable by un-trimming.
**Peter 05Oct23 testing with latest 4 of 6 branch:**    I can't see how to do the paste-over, so can't test this.
- [x] "Silence Audio" command over several tracks and stretched clips shows a convincing progress bar
**Peter 05Oct23 testing with latest 4 of 6 branch:**   I get a "convincing progress bar" with both stretched and unstretched clips - each clip was a one hour chirp and I had four of them.  The silent audio was properly rendered to 100%
- [x] Same for the AutoDuck effect, with stretch in the control track.
**Peter 05Oct23 testing with latest 4 of 6 branch:**   I get a "convincing progress bar" with both stretched and unstretched clips - each clip was a one hour chirp and I had four of them.  The ducked audio was properly rendered to 100%
- [x] Equalization effects (such as Graphic EQ) on unstretched tracks have the same behavior as before (i.e., merging begin and end of output when possible, preserving clip boundaries within the selection, etc.)
**Peter 05Oct23 testing with latest 4 of 6 branch:**   Seems to work properly. But if I have a gap between clips I get an unwanted additional silent clip between the separated clips (which of course impacts on clip movements)
- [x] Verify that this PR fixes #5265
**Peter 05Oct23 testing with latest 4 of 6 branch:**    This looks to work properly in the branch
- [x] Check if this PR fixes #4250
**Peter 05Oct23 testing with latest 4 of 6 branch:**    This looks to work properly in the branch
- [x] [Bug found by @Tantacrul on master](https://www.dropbox.com/s/0hlb011cx0rx4wr/Pasting%20Mutliple%20clips%20next%20to%20other%20clips.mov?dl=0) is fixed
- [x] Implements #5273
**Peter 05Oct23 testing with latest 4 of 6 branch:**    This renders the clips but I get glitches/clicks - see bug thread for details
- [x] Insert silence on stretched clip with non-trivial envelope behaves as expected
- [x] Cutline-cutting stretched region with envelope and expanding back restores the envelope
- [x] Paulstretch on both unstretched and stretched clip
**Peter 05Oct23 testing with latest 4 of 6 branch:**    seems to work fine with both stretched and un-stretched.

Paul-Licameli commented:
The first commit in this branch fails to build.

saintmatthieu commented:
> The first commit in this branch fails to build.

Squashed the branch to something more sensible. Each commit builds now.

Paul-Licameli commented on src/tracks/playabletrack/wavetrack/ui/CutlineHandle.cpp:
```diff
@@ -169,7 +169,16 @@ UIHandle::Result CutlineHandle::Click
-      else if (mLocation.typ == WaveTrackLocation::locationMergePoint) {
+      // todo(mhodgkinson) Bluntly disabling this Join shortcut for now. There
+      // exists an issue that wants to do this properly:
+      // https://github.com/audacity/audacity/issues/5020
```
Replicated in another PR -- again, wrong number?

saintmatthieu replied:
Spot on.

Paul-Licameli commented on libraries/lib-track/Track.cpp:
```diff
@@ -51,7 +51,7 @@ Track::Track(const Track& orig, ProtectedCreationArg&&)
 {
    mId = orig.mId;
-
+   mProjectTempo = orig.mProjectTempo;
```
Please edit the commit comment.  Put breaks in the very long line that is the third paragraph, and remove the last line.

saintmatthieu replied:
I reverted this change - wrong since `mProjectTempo` became part of `ChannelGroupData`.

Paul-Licameli commented on src/menus/EditMenus.cpp:
```diff
@@ -138,9 +139,16 @@ void DoPasteNothingSelected(AudacityProject &project, const TrackList& src, doub
+   const double srcTempo =
+      pFirstNewTrack ? pFirstNewTrack->GetProjectTempo().value_or(projTempo) :
+                       projTempo;
+   // Apply adequat stretching to the selection. A selection of 10 seconds of
```
"Adequate"

Paul-Licameli commented on src/menus/EditMenus.cpp:
```diff
@@ -138,9 +139,16 @@ void DoPasteNothingSelected(AudacityProject &project, const TrackList& src, doub
+   const double srcTempo =
+      pFirstNewTrack ? pFirstNewTrack->GetProjectTempo().value_or(projTempo) :
+                       projTempo;
+   // Apply adequat stretching to the selection. A selection of 10 seconds of
```
A test of the branch tells me I have described how it behaves.  Whether that is as design specified, I don't know, but assuming so, I update the QA checklist.

saintmatthieu replied:
Yes, that's correct. It's a simple UX rule: preserve the length in beats rather than in seconds.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -610,6 +637,7 @@ bool WaveClip::Paste(double t0, const WaveClip &other)
 
        auto copy = std::make_shared<WaveClip>(other, factory, true);
        copy->ClearSequence(copy->GetSequenceStartTime(), copy->GetPlayStartTime());
+       copy->SetTrimLeft(0);
```
Does this added line, with the other at 630, fix a known bug in copy-paste?  Strictly speaking that's not the scope of the PR or the stretching project.  Is there an issue number that it fixes?

Paul-Licameli replied:
I will add to the QA checklist anyway

saintmatthieu replied:
The reason is obscure, but there is one.
`copy->ClearSequence(playEndTime, sequenceEndTime)` clears the trimmed data. If we don't call `copy->SetTrimLeft(0)`, one may still think the hidden data is valid.
Further down I call `newClip->ApplyStretchRatio()`. This provisionally sets left and right trim to 0, to reduce the ramping in/out of the effect if possible ...
Note that, in 5-of-6, I remove that call on `newClip`, but this is a healthy change in any case, making changes in the lines below safer.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -625,13 +653,19 @@ bool WaveClip::Paste(double t0, const WaveClip &other)
       newClip = std::move(copy);
    }
+   ApplyStretchRatio();
+   newClip->ApplyStretchRatio();
```
The function might be more understandable if you do that stretch only when `!stretchOtherBeforeResampling`

But I understand that if `stretchOtherBeforeResampling`, then the line does nothing because the stretch ration is already unit.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -1099,6 +1129,71 @@ void WaveClip::Resample(int rate, BasicUI::ProgressDialog *progress)
+   SetTrimRight(0);
+
+   constexpr auto durationToDiscard = 0.0;
+   constexpr auto blockSize = 1024;
```
Did experimentation with larger numbers might find an easy way to improve performance?

Paul-Licameli reviewed (changes requested):
Please complete the fixes for exception safety.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -358,6 +358,19 @@ void WaveTrack::DoOnProjectTempoChange(
+   const auto originalTempo = other.GetProjectTempo();
+   const auto provisionalTempo = GetProjectTempo();
+   if (!provisionalTempo.has_value() || !originalTempo.has_value())
+      THROW_INCONSISTENCY_EXCEPTION;
```
Why this exception?

Paul-Licameli replied:
I wonder if instead this should be an assertion that is stated as a precondition.  But then the precondition may need to be repeated for the functions that call this, and so on up the stack until somewhere you can give some convincing argument that the condition is guaranteed.

saintmatthieu replied:
Now I think this throw behaviour is too strict. This method is used so far in `WaveTrack::PasteWaveTrack` and `WaveTrack::ClearAndPaste`. I cannot think of a reason for either of them not to be callable before they are made part of a project, but maybe I'm not thinking hard enough.
A more pragmatic approach would be for this method to assert and return without doing anything instead of throwing, without stating a pre-condition. The assert might reveal a concrete example where this doesn't hold in the future, yet there might not be any consequence at all (e.g. if no stretching occurred), in which case bothering the user with it would just be annoying to everyone.
Ok for assert without pre-condition ?

Paul-Licameli replied:
Now I think, let the exception stand.  Rewrite the function as I suggested so it doesn't mutate `other` but constructs a copy.

If the preconditions are not satisfied, throw.  The two places where this function is called are in pasting and I already made sure years ago with exhaustive examination of call paths that these are safe places to throw and have `AudacityApp::OnExceptionInMainLoop` catch it and deal with it.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -358,6 +358,19 @@ void WaveTrack::DoOnProjectTempoChange(
+      THROW_INCONSISTENCY_EXCEPTION;
+   other.OnProjectTempoChange(*provisionalTempo);
+   return { [originalTempo, &other] {
+      other.OnProjectTempoChange(*originalTempo);
```
1. What happens inside a `Finally` may happen during stack unwinding, so the called functions should be `noexcept`.  (I don't believe in adding `noexcept` just because you can; but only when there are certain strong reasons for it like here.)

2. This might not restore the state of the other track exactly.  `mRawAudioTempo` in a clip may go from empty to non-empty and not back again.

Paul-Licameli replied:
The two places that use this function also use `const_cast`.  I feel some unease.  I might approve reluctantly.

saintmatthieu replied:
>  the called functions should be `noexcept`

I never used that keyword before. I notice, though, that
```
void foo() { throw std::runtime_exception(""); }
void bar() noexcept { foo(); }
```
doesn't yield a warning, which would make tedious checking that all nodes of the call tree of `OnProjectTempoChange` also are `noexcept`. Or am I missing something ?

saintmatthieu replied:
> This might not restore the state of the other track exactly.

I hadn't thought about that, but looks like I'm a lucky guy: `track.GetProjectTempo().has_value()` if and only if `track.OnProjectTempoChange` was called at least once. If it wasn't, then this method doesn't proceed. If it was, then all clips have `mRawAudioTempo` set.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1477,166 +1319,158 @@ void WaveTrack::PasteWaveTrack(double t0, const WaveTrack &other)
+   }
+   }
+
+   assert(false);
```
Move this to a `default:` case of the `switch` to remove a compile warning.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -861,6 +874,11 @@ auto WaveTrack::CopyOne(
+TrackListHolder WaveTrack::Copy(bool forClipboard) const
+{
+   return Copy(GetStartTime(), GetEndTime(), forClipboard);
+}
```
This overload is not used

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2658,15 +2496,9 @@ const WaveClip* WaveTrack::GetAdjacentClip(
-                   1e-9 ?
-                neighbour :
-                nullptr;
+      return neighbour->SharesBoundaryWithNextClip(&clip) ? neighbour : nullptr;
```
The definition of this function is better now, but it is still not used

saintmatthieu commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -1099,6 +1129,71 @@ void WaveClip::Resample(int rate, BasicUI::ProgressDialog *progress)
+         assert(this->GetStretchRatio() == 1.0);
+      }
+      else
+         std::swap(mSequences, newSequences);
```
Agreed. Cherry-picked your fix-commit d15f03a1c4ca0b6f59d286ac59862635b3d27457.

saintmatthieu commented:
I reset this PR to draft because I have to make a bigger effort at preserving the old behaviour where appropriate. Will update the QA checklist for use are reference.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -944,6 +944,12 @@ class WAVE_TRACK_API WaveTrack final
       const std::optional<double>& oldTempo, double newTempo) override;
+   // If `other` may come from another project with a different tempo, its tempo
+   // must be set to the current project's to get correct timing of audio clips.
+   // Careful: `other` must outlive the score where you store the returned
```
"scope"

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1135,6 +1148,12 @@ void WaveTrack::ClearAndPaste(double t0, // Start of time to clear
+   // This has to be before `src.GetStartTime()` and such, or the boundaries
+   // won't be correct.
+   const auto finally =
+      ProvisionallySynchronizeProjectTempo(const_cast<WaveTrack&>(src));
```
I still dislike the `const_cast` in two places and the non-obvious reasoning why there is never a net side effect on `WaveClip`.

An alternative is that `ProvisionallySynchronizeProjectTempo` (renamed maybe) takes a reference to a `const WaveTrack` and returns a shared pointer to a new, duplicate WaveTrack that has the project tempo change in it.  Its scope is only as long as this `finally` would be.

Maybe a lot of memory allocations and deallocations but a sharing of all the blocks in the Sqlite database so none of the very expensive things are done.

Paul-Licameli replied:
Consider too that duplication of the in-memory WaveTrack structures, but with sharing of Sqlite resources, happens all the time when you push and pop the Undo stack, or even just modify the current entry because of a change of the selected interval.  And we hardly notice the performance.

saintmatthieu replied:
That's a good idea, better than this kind of magic I was doing.
See 8604d5726a239849bdd80636a3d948c31e076ccd.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -535,6 +535,19 @@ void WaveTrack::DoOnProjectTempoChange(
+   const auto originalTempo = other.GetProjectTempo();
+   const auto provisionalTempo = GetProjectTempo();
+   if (!provisionalTempo.has_value() || !originalTempo.has_value())
+      THROW_INCONSISTENCY_EXCEPTION;
```
Now I think, let the exception stand. Rewrite the function as I suggested so it doesn't mutate other but constructs a copy.

If the preconditions are not satisfied, throw. The two places where this function is called are in pasting and I already made sure years ago with exhaustive examination of call paths that these are safe places to throw and have AudacityApp::OnExceptionInMainLoop catch it and deal with it.

Paul-Licameli commented on src/menus/EditMenus.cpp:
```diff
@@ -138,9 +139,16 @@ void DoPasteNothingSelected(AudacityProject &project, const TrackList& src, doub
+   // audio in project A should become 5 seconds in project B if tempo in B is
+   // twice as fast.
+   const double quantT0 = QUANTIZED_TIME(t0 * srcTempo / projTempo, projRate);
+   const double quantT1 = QUANTIZED_TIME(t1 * srcTempo / projTempo, projRate);
```
The commit comment "Preserve beat length of selection when copying across projects" describes this part of it.  Perhaps the rest of this commit needs other commit comments, whether in the same or a separate commit.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -226,11 +226,11 @@ void WaveClip::OnProjectTempoChange(
       // the meantime, use the tempo of the project when the clip is created as
       // source tempo.
-      mRawAudioTempo = oldTempo;
+      mRawAudioTempo = oldTempo.value_or(newTempo);
```
I don't remember what this line fixes -- was it only necessary when you were using temporary side effects on clips?

saintmatthieu replied:
Not sure anymore either, but it still is better this way: `WaveClip::OnProjectTempoChange` is the only way a clip's project tempo can be set, hence `WaveTrack::InsertClip` calling it with `oldTempo == nullopt`.
Best would be to add a dedicated API, but I'm lazy.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -226,11 +226,11 @@ void WaveClip::OnProjectTempoChange(
    if (oldTempo.has_value())
    {
-      const auto ratioChange = *mProjectTempo / newTempo;
+      const auto ratioChange = *oldTempo / newTempo;
```
But this change makes sense regardless

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.h:
```diff
@@ -957,6 +965,13 @@ class WAVE_TRACK_API WaveTrack final
+   /*
+    * @copybrief ClearAndPasteAtSameTempo
+    * @pre Preconditions of `PasteWaveTrack`
+    * @ore `GetProjectTempo().has_value() && GetProjectTempo() ==
```
`@pre`

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2969,8 +2969,7 @@ const WaveClip* WaveTrack::GetClipAtTime(double time) const
       clips.rbegin(), clips.rend(), [&](const WaveClip* const& clip) {
-         return time >= clip->GetPlayStartTime() &&
-                time <= clip->GetPlayEndTime();
+         return clip->WithinPlayRegion(time);
```
A small change.  But I remember some of the surprises in getting effects figured out, were the unintended consequences of changing `<` to `<=` tests in certain `WaveTrack` and `WaveClip` member functions that were more widely used.

Are there such surprises here?

I did review all uses of `GetClipTime` and now, I think not.

saintmatthieu replied:
When I changed the boundary check in the first place, it was to redefine clip intervals as standard closed-begin, open-end. Consistently, and for adjacent clips, `clipA.GetPlayEndTime() == clipB.GetPlayEndTime()`.

This is a place that went against this rule of thumb that I haven't addressed earlier, and would prevent me from using `GetClipAtTime` in `PasteOne`: there I don't want a clip whose end is at `t0` to interfere.

I also looked at uses of `GetClipAtTime`. It's used in `GetEnvelopeAtTime`. My testing of envelope editing showed that it still worked properly. Will add a QA checkbox.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -3199,7 +3199,7 @@ void WaveTrack::SplitAt(double t)
    for (const auto &c : mClips)
    {
-      if (c->WithinPlayRegion(t))
+      if (c->SplitsPlayRegion(t))
```
This simple change raises similar doubts.  But I presume you have retested the destructive effects with this change?  This PR is based on all of that other stuff.

I have in fact noticed zero-length clips sometimes as a thing that ought to be debugged later.

Is there a way of making zero length clips in 3.3.3 that gets fixed by this?

Maybe there are existing issues about them?

saintmatthieu replied:
Not that I know of. #2263 wrongly reports zero-length clips, but that's actually the click-to-merge vertical line.
In any case, it'd be wrong indeed to split a track on the first sample of a clip.
One could even question whether a track should be splittable at the second-last sample of a clip, leaving a length-1 clip, barely more usable than a length-0 clip. But I didn't venture that far.

Paul-Licameli commented on src/menus/LabelMenus.cpp:
```diff
@@ -182,18 +187,48 @@ void EditByLabel(AudacityProject &project,
+      const bool playable = dynamic_cast<const PlayableTrack*>(t) != nullptr;
+      if (SyncLock::IsSyncLockSelected(t) || (notLocked && playable))
+         tracksToEdit.push_back(t);
    }
```
I might write this without the vector:

`const auto range = tracks.Any<PlayableTrack>() + [&](const auto pTrack){ return SyncLock::IsSyncLockSelected(t) || (notLocked && playable); }`

Paul-Licameli replied:
Then just use `range` twice in range-for statements

saintmatthieu replied:
IIUC, the `+` overload works like an "&&", and your proposal would yield another condition: `playable && (IsSynCLockSelected || notLocked)`.
I think what could work is
```
const auto range = tracks.Any<Track>() + [&](const auto pTrack) {
      return SyncLock::IsSyncLockSelected(t) ||
             (notLocked && dynamic_cast<const PlayableTrack*>(pTrack) != nullptr);
   };
```
but that's not so important.

Paul-Licameli commented on src/menus/LabelMenus.cpp:
```diff
@@ -591,11 +626,15 @@ void OnJoinLabels(const CommandContext &context)
+   EditByLabel(
+      project, tracks, selectedRegion, editfunc,
+      BasicUI::MakeProgress(
+         XO("Pre-processing"), XO("Rendering Time-Stretched Audio")));
```
Strings.  @LWinterberg  ?

saintmatthieu replied:
Part of the utility functions proposed in recent commits.

Paul-Licameli commented on src/menus/EditMenus.cpp:
```diff
@@ -898,9 +898,18 @@ void OnJoin(const CommandContext &context)
-   for (auto wt : tracks.Selected<WaveTrack>())
-      wt->Join(selectedRegion.t0(), selectedRegion.t1());
+   const auto progress = BasicUI::MakeProgress(
+      XO("Pre-processing"), XO("Rendering Time-Stretched Audio"), 0);
```
I see a repetition of these strings and would like to see a convenience function to make the progress indicator, so the strings occur in just one place

saintmatthieu replied:
Addressed - see proposal in more recent commits.

Paul-Licameli commented on libraries/lib-wave-track/PasteOverPreservingClips.cpp:
```diff
@@ -76,21 +76,23 @@ void PasteOverPreservingClips(const ClipData &data,
-      if ((realStart  != start || realEnd != end) &&
-         !(realStart <= startT && realEnd >= startT + lenT))
-         oldTrack.Join(realStart, realEnd);
+      oldTrack.Join(
```
A conditional call to `Join` is now unconditional.  Does this change behavior?

In the case that the selected portion of track for EQ contains split lines, not at the extremes of the selection, then this is intended to put the lines back where they were.  Does this conserve that behavior?

saintmatthieu replied:
I think that `clipRealStartEndTimes` contain the original clip boundaries, and `Join` isn't called across any of these.

But I'm trying to figure out if this function is needed at all. What it does:
Takes `newContents`, a one-clip track with the output of the effect, replaces the content of `oldTrack` in `region = [start, start + len)` with it, preserving existing clip boundaries strictly within `region`, but merging at `start` if it splits a clip underneath, and likewise for `start + len`.
This is the intended behaviour for other effects - if merging doesn't mean stretch-rendering, else there is no merging. Can you recall why this is a special case ?

Also, it crashes on `master` for [this project](https://github.com/audacity/audacity/files/12657643/test.zip): not modifying the selected region, apply some eq. `GetClipAtTime` above returns an unchecked nullptr. I haven't got to the bottom of this but if this function isn't needed I may not have to. #5138 doesn't repair it but this branch does somehow.

Paul-Licameli commented:
Please review and approve this new commit https://github.com/audacity/audacity/pull/5138/commits/1372b859c9c2db198da93129729d8fba0f0371a6 at #5138 

-- and then make similar fixes for this branch

Paul-Licameli commented:
Added a QA checkbox about sync lock

saintmatthieu commented:
@Paul-Licameli addressed your comments, with [this](https://github.com/audacity/audacity/pull/5043#discussion_r1322887539) as yet unresolved question.

Paul-Licameli commented:
I am beginning another round of review, checking that each commit builds.

The first fails to build but it is easily fixed.

Paul-Licameli commented on libraries/lib-effects/EffectOutputTracks.cpp:
```diff
@@ -50,7 +50,8 @@ EffectOutputTracks::EffectOutputTracks(
       auto progress = MakeProgress(
-         XO("Pre-processing"), XO("Rendering Time-Stretched Audio"),
+         WaveTrackUtilities::stretchRenderingProgressBarTitle,
+         WaveTrackUtilities::stretchRenderingProgressBarMessage,
```
This commit eliminates some duplication.  But it doesn't do enough.

See this in `EffectOutputTracks.cpp`:

```
      auto reportProgress = [&](double progressFraction) {
         const auto overallProgress = (count + progressFraction) / numTracks;
         const auto result = progress->Poll(overallProgress * 1000, 1000);
         if (result != ProgressResult::Success)
            throw UserException {};
      };
```

This lambda needs to be defined only once, and reusable.  In the other two places that use the strings, they repeat part of this lambda, but not the reset of it -- therefore those progress indicators have non-functional Cancel buttons!

The utility should take a reference to `count` and construct a lambda that captures it.  The utility returns a `std::function` and comments that it may throw `UserException`.

Paul-Licameli replied:
There is a fourth place -- in AutoDuck.  The identical lambda could be eliminated.

The utility could construct the progress indicator.  Do you know C++14 extended lambda capture?  It lets a lambda hold a `unique_ptr` which C++11 could not do.

```
[
   progress = MakeProgress(...),
  ...
]
```

There is need to vary the progress bar message for AutoDuck but nothing else.  So the utility could take that as another parameter.

Paul-Licameli replied:
You would need another parameter too, the denominator of the division.

Paul-Licameli commented on src/menus/LabelMenus.cpp:
```diff
@@ -177,13 +177,17 @@ using EditFunctionWithProgress =
+   // Apply action on tracks starting from
+   // labeled regions in the end. This is to correctly perform
+   // actions like 'Delete' which collapse the track area.
+   std::reverse(regions.begin(), regions.end());
```
Okay, this commit totally rewrites the previous.  But it does it with a very weird and unnatural inversion of control.  What led to this choice?

I think it would make more understandable code if instead you had a factory making a function object that you just call to update the user, and you write loops more naturally, just bumping the count captured by reference in the factory.

Did you not do this just because you could not figure out the lambda capture?

Abstracting just the test for user cancellation means decisions about how to iterate the range aren't forced -- and you would not need the line above but could iterate with `rbegin` and `rend` as it was.

Really I don't like this.  Try it as I suggested.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1939,19 +1941,11 @@ void WaveTrack::ApplyStretchRatio(
+      reportProgress, TrackList::Channels(this),
+      [&](WaveTrack* pChannel, ProgressReporter child) {
+         pChannel->ApplyStretchRatioOne(startTime, endTime, child);
+      });
```
This commit reuses the utility, for nested loops.

Maybe I understand better now why that led to the strange inverted flow of control.

Still I don't like it.  The simpler alternative to eliminate repetition would be to write another utility that wraps one progress indicator in another, that takes another denominator and another reference to a count.

Paul-Licameli replied:
Yes like that BUT — the lambda must capture with = not & or you get a dangling reference!

saintmatthieu replied:
Following-up our discussion, I tried your idea to get inspiration from `std::for_each` - _that_ is the kind of suggestion I was hoping you were going to make from the beginning. Works like a charm, see [Morre standard-conform SplitProgress](https://github.com/audacity/audacity/pull/5043/commits/3d0ba5ae52150a11564272ab476aec01bc4c937d).

Paul-Licameli commented on src/menus/EditMenus.cpp:
```diff
@@ -615,8 +615,15 @@ void OnPaste(const CommandContext &context)
+                  constexpr auto preserveExistingBoundaries = false;
+                  wn.ClearAndPaste(
+                     t0, t1, *static_cast<const WaveTrack*>(src),
+                     preserveExistingBoundaries, merge, &warper);
```
This commit is small, but I remember now the doubts I had about it.  The changes in Generator led me here, and just that led me to separate the destructive effects branch first, and get all of its problems figured out first.

My doubt is because is that the boolean `merge` governs more than just the internal splitting of the new pasted clip:  see all the rest in `ClearAndPasteOne` under `if (merge && splits.size() > 0)`

I was thinking then, and write it down now, that I think the correct fix may require more boolean arguments to control the details more finely.

Paul-Licameli replied:
Contrast the behavior:
- Select part of a track, with a stretch.  Apply an effect. Drag the new clip to another track.  Drag (don't stretch) clip edges.
- Similar but pasting instead of applying an effect (and there doesn't need to be any stretch).

For the effect, the clips containing unchanged data retain hidden clip information.  But for paste -- as a consequence of this commit -- they do not.  (Without this commit, of course there were no new splits at the selection bounds.)

Which is the correct behavior?  I think there should be consistency between these cases.

As for the new clip in the middle -- for the effect, it turns out to preserve the rest of the clip, stretched but without the effect applied.  (But #5220 will change that behavior for performance reasons).

For paste, the middle clip can still carry hidden data from the place from which it was originally cut or copied.

saintmatthieu replied:
Great catch, thank you.
We could add one more boolean, to control the `split` argument of the internal call to `HandleClear`, currently hard-coded to false. This would solve the problem you describe, but would introduce new problems: "There is not enough room ..." if the pasted data is larger than `[t0, t1]`, and a unfilled gap if the pasted data is less than `[t0, t1]`.
And me believing I was done with it ...

saintmatthieu replied:
[Generalize ClearAndPaste to allow preservation of cleared interval](https://github.com/audacity/audacity/pull/5043/commits/fbfbbf5f5d1ab873db6eec0437a6130a0def8f41)

Paul-Licameli commented:
> IIUC, the `+` overload works like an "&&", and your proposal would yield another condition: `playable && (IsSynCLockSelected || notLocked)`. I think what could work is
> 
> ```
> const auto range = tracks.Any<Track>() + [&](const auto pTrack) {
>       return SyncLock::IsSyncLockSelected(t) ||
>              (notLocked && dynamic_cast<const PlayableTrack*>(pTrack) != nullptr);
>    };
> ```
> 
> but that's not so important.

Please reconsider that, I find it more elegant to read it.

Paul-Licameli commented:
My review so far.

The last things I don't fully understand are the implications of the changes in `PasteOverPreservingClips.cpp` and the changes in `WaveTrack::PasteOne`.

About the changes to remove duplication with progress indicators -- you can just remove those three commits and leave the duplication of code for now.  We can make the code nicer later.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1758,134 +1805,63 @@ void WaveTrack::PasteOne(
-    //wxPrintf("paste: we have at least one clip\n");
-
-    bool singleClipMode = other.GetNumClips() == 1 &&
-        std::abs(startTime) < track.LongSamplesToTime(1) * 0.5;
```
The argument `startTime` is now unused, so remove it

saintmatthieu replied:
... which makes me realize that, in `WaveTrack::PasteWaveTrack` and `ClearAndPasteOne`, the `endTime` argument was wrongly passed as `insertDuration` argument to `PasteOne`. I suppose this wasn't noticed by QA because these two outer methods are called with temporary tracks, all with their first clip at 0.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1758,134 +1805,63 @@ void WaveTrack::PasteOne(
+
+   // Make room for the pasted data
+   track.SplitAt(t0);
+   const auto sortedClips = track.SortedClipArray();
```
This needs scope only within the `else` block

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1758,134 +1805,63 @@ void WaveTrack::PasteOne(
+      auto clipAtT0 = track.GetClipAtTime(t0);
+      const auto firstClipNotToShift = std::find_if(
+         sortedClips.begin(), sortedClips.end(),
+         [&](const WaveClip* clip) { return clip->BeforePlayRegion(t0); });
```
I might rewrite this with `std::lower_bound`

dozzzzer commented:
A note for myself: check if the resulting clip gets shifted a few samples rightwards after having merged the two stretched adjacent clips:

https://github.com/audacity/audacity/assets/70063970/025fe2d4-2190-4f74-9eff-8dd39cb1113f

saintmatthieu commented:
There's a crash when joining clips with contrasting stretch ratios. I verified that the [More standard-conform SplitProgress](https://github.com/audacity/audacity/pull/5043/commits/3d0ba5ae52150a11564272ab476aec01bc4c937d) commit isn't responsible, but couldn't investigate any further at this time.

Paul-Licameli commented on libraries/lib-basic-ui/BasicUI.h:
```diff
@@ -315,20 +315,24 @@ inline std::unique_ptr<GenericProgressDialog> MakeGenericProgress(
+      action(*first, child);
       ++count;
    }
 }
```
Now the utility function is nicer to use, but I feel that BasicUI.h is the wrong place for it.  I'm not sure what is the right place for it.

Paul-Licameli replied:
The loops and nested loops surely look much nicer now.

I think the function needs a better name than `SplitProgress`.  Perhaps `WithRedenominatedProgress` and some comments explaining the change of denominator as the size of the given iterator range.

saintmatthieu replied:
I made several attempts at finding a better name, but then came to think that it wasn't such a bad name after all.
Added a description in 1bc93b7.

Paul-Licameli commented on libraries/lib-basic-ui/BasicUI.h:
```diff
@@ -315,20 +315,24 @@ inline std::unique_ptr<GenericProgressDialog> MakeGenericProgress(
+   ItType first, ItType last,
+   std::function<void(typename ItType::value_type, const ProgressReporter&)>
+      action,
+   ProgressReporter parent)
```
`ProgressReporter` too can be a deduced type, allowing more generality and avoiding some more `std::function` constructions.

saintmatthieu replied:
But doesn't the user then loose the information that it takes a double and returns a void just by looking at the function's signature?

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1991,12 +1993,13 @@ void WaveTrack::ApplyStretchRatioOne(
       interesectedClips.push_back(clip);
```
"intersected" typo

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1417,16 +1417,21 @@ void WaveTrack::ClearAndPasteOne(
+         // `src` was created by copy from `target`, so they have equal width
+         // and stretch ratio.
+         assert(target.GetWidth() == src.GetWidth());
+         assert(target.GetStretchRatio() == src.GetStretchRatio());
```
Not just commenting assertions but adding one.  So can you prove this new one is satisfied at the point of the call?

saintmatthieu replied:
Provided what the comment I added is correct, yes.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2308,28 +2392,33 @@ void WaveTrack::JoinOne(WaveTrack &track, double t0, double t1)
       //wxPrintf("Pasting at %.6f\n", t);
-      bool success = newClip->Paste(t, *clip);
+      bool success = newClip->Paste(t, *clip, reportClipProgress);
       assert(success); // promise of CreateClip
```
Indentation

Paul-Licameli commented on libraries/lib-wave-track/WaveTrackUtilities.cpp:
```diff
@@ -23,3 +28,18 @@ bool WaveTrackUtilities::HasStretch(
 }
+
+void WaveTrackUtilities::WithStretchRenderingProgress(
+   std::function<void(const ProgressReporter&)> action,TranslatableString title,
```
Space after ,

Paul-Licameli commented:
The commit "WaveClip::Paste handles stretch ratio differences" no longer builds.  This call has too few arguments:

```
      GetClip(channel)->ApplyStretchRatio(
         [&reportProgress, channel, channelsCount](double progress)
         { reportProgress((channel + progress) / channelsCount); });
```

Paul-Licameli commented:
Approved for QA testing, but I want to continue with my own open-box testing too, exploring edge cases

Paul-Licameli commented:
I see the additional commit, and it mostly looks good, but one question...

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1987,6 +1986,14 @@ void WaveTrack::PasteOne(
+       if (!track.IsEmpty(t, t + insertDuration))
+          throw notEnoughSpaceException;
+       track.SplitAt(t0);
+       clipAtT0->ShiftBy(insertDuration);
```
Should you reassign `clipAtT0` after `SplitAt`?  To be sure you shift the correct one and make no assumptions about how `SplitAt` mutates the clips previously present?

saintmatthieu replied:
Yes, that's a potential crash indeed. Will publish a fix ...

dozzzzer commented:
Found issues @saintmatthieu 

- when joining 3 or more clips having different stretch ratios, the resulting clips has wrong length and has gaps of silence that aren't supposed to be there:

https://github.com/audacity/audacity/assets/70063970/72dfc1c5-81c0-40eb-a59f-68ced3ea9a4a

- when having multiple clips with different stretch ratios across multiple tracks, selecting them all and trying to Join, Audacity freezes. [This project ](https://www.dropbox.com/scl/fi/0sziskbfz13g35uhjyrhl/buggy-proj.aup3?rlkey=fwls04x70ebv8dk6x8m7myat4&dl=0)can be used for reproduction.

https://github.com/audacity/audacity/assets/70063970/3679e8b2-d1aa-4d75-806e-97056a77dbef

(to be continued)

saintmatthieu commented:
I realize that the commit I just pushed, [Fix potential crash when joining stereo clips](https://github.com/audacity/audacity/pull/5043/commits/fe8331d3bd3089c2c346a0f21352f59a3596bf87), while fixing the problem when joining two clips, is insufficient when joining more ...

saintmatthieu commented:
> * when having multiple clips with different stretch ratios across multiple tracks, selecting them all and trying to Join, Audacity freezes. [This project ](https://www.dropbox.com/scl/fi/0sziskbfz13g35uhjyrhl/buggy-proj.aup3?rlkey=fwls04x70ebv8dk6x8m7myat4&dl=0)can be used for reproduction.
> 
>  Screen.Recording.2023-09-28.at.17.58.04.video-converter.com.mp4 

Strange that some part of some of your clips don't have the usual darker background, isn't it? Do you know how it came to this? 
![image](https://github.com/audacity/audacity/assets/22740106/d9648001-4be6-43d1-b4bd-58f7216cc69c)

chinakov commented:
Bug found by @Tantacrul is fixed, thank you! Reflected in the checklist.

dozzzzer commented:
@saintmatthieu In the updated branch, previously logged issues are now fixed 👍 
New issues found:

### When generating audio into a clip, the resulting clip is shifted left to zero and the waveform is incorrect:

#### Expected:

https://github.com/audacity/audacity/assets/70063970/151c7d1f-7314-48f2-abb6-28f417d5821c

#### Actual:

https://github.com/audacity/audacity/assets/70063970/732c98eb-7a7c-4f1b-a88e-93ddcdb4fbf6


### When joining two stretched clips with envelopes, the resulting clip's amplitudes don't match the pre-rendered ones:


https://github.com/audacity/audacity/assets/70063970/8dc2f2c0-b16f-4b9f-8e92-daf8647d7094

#### Before `Audio Clips > Join`:

<img width="450" alt="Screenshot 2023-10-02 at 18 34 34" src="https://github.com/audacity/audacity/assets/70063970/0ceda231-f803-4172-8897-10293d11dfe9">

#### After `Audio Clips > Join`:

<img width="450" alt="Screenshot 2023-10-02 at 18 34 50" src="https://github.com/audacity/audacity/assets/70063970/1870a4c1-2afe-4bd5-b96c-79f1d578d0ef">

(to be continued)

saintmatthieu commented:
Thanks @dozzzzer.
I noticed yet other problems: at the end of a clip,

![image](https://github.com/audacity/audacity/assets/22740106/3f6212cb-8383-4010-9c8d-38384ddd3a6b)

generate a chirp. You get this ugly result:

![image](https://github.com/audacity/audacity/assets/22740106/722a5e14-39e2-44ae-8734-2b589cdac7ba)

Unfortunately I didn't manage solving this before my day off tomorrow. @Paul-Licameli if you have the capacity to look into this. Clue: the `PasteTimeWarper` is causing problems when restoring the splits in `ClearAndPasteOne`.

dozzzzer commented:
@saintmatthieu in the updated PR:

- the issue with generating audio at the end of the clip (see previous comment) is now fixed
- the issue with generating audio over a stretched clip is now fixed
- the issue with joining clips with different stretched ratios and envelopes remains **unfixed**
- there's another issue with generating audio (only in case of Nyquist generators, such as e.g. Pluck) over an interval selected on a stretched clip:

https://github.com/audacity/audacity/assets/70063970/4190430c-cb00-4057-964d-9a4b81678c87

saintmatthieu commented:
> * the issue with joining clips with different stretched ratios and envelopes remains **unfixed**

Sorry I oversaw that. Commits between your comment and now address this.

saintmatthieu commented:
> there's another issue with generating audio (only in case of Nyquist generators, such as e.g. Pluck) over an interval selected on a stretched clip

Actually I also hit some assert if the length of generated audio is greater than the selection - doesn't have to be Nyquist. (Don't know how it looks in a Release build.) Looking into this ...

Paul-Licameli commented:
Intermediate step “WaveClip::Paste handles stretch ratio differences” now fails to build

Paul-Licameli commented:
The build failure is fixed three commits later

saintmatthieu commented:
> there's another issue with generating audio (only in case of Nyquist generators, such as e.g. Pluck) over an interval selected on a stretched clip:

I found the issue, which indeed didn't concern Nyquist only but also some generators (e.g. tone generator). Generating audio over a selection for a duration greater than the selection was buggy. (Haven't checked if it was also buggy for durations less than the selection.)

With the tone generator, the duration is set per default to the selection duration, but not with the Nyquist generator, and that got you to notice the bug.

For what it's worth, the fix didn't involve modifying that complicated `ClearAndPaste` method, but usages of it, so I'm optimistic that this change won't break other things.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -759,9 +760,11 @@ bool WaveClip::Paste(
+   const auto timeOffsetInEnvelope =
+      s0.as_double() * thisStretchRatio / mRate + GetSequenceStartTime();
+   mEnvelope->PasteEnvelope(
+      timeOffsetInEnvelope, newClip->mEnvelope.get(), sampleTime);
```
I see the commit that fixes envelopes.  I read through all uses of envelopes in this file, to see if it is complete.

Do these need any changes?  I think not, but let me know the code paths were tested too and recommended to QA.

`InsertSilence` 

`AppendSilence`

`ClearAndAddCutLine`

`ExpandCutLine`

saintmatthieu replied:
Not tested but examined: the other places pass unstretched time values to the envelope methods, which looks correct. I added QA test cases nevertheless 👍

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -1439,17 +1439,20 @@ void WaveClip::TrimRight(double deltaTime)
 double WaveClip::GetSequenceStartTime() const noexcept
 {
     // JS: mSequenceOffset is the minimum value and it is returned; no clipping to 0
+    // Do we need to `SnapToTrackSample` before returning?
```
This comment raises a lot of questions about a function used in many places.  Did you study all those uses?

saintmatthieu replied:
No I haven't.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -280,6 +280,13 @@ double WaveTrack::Interval::GetPlayEndTime() const
+bool WaveTrack::Interval::IntersectsPlayRegion(double t0, double t1) const
+{
+   // TODO wide wave tracks:  assuming that all 'narrow' clips share common
+   // boundaries
```
You don't need the comment above.  We have done the work to guarantee it now that this is always so: upgrading the opening of old files with misalignment, and changing what happens when mono tracks are joined into stereo.

saintmatthieu replied:
That doesn't invalidate this statement that we are assuming equal boundaries, though, does it?

Paul-Licameli replied:
We do assume that, and there will be nothing left TODO later

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2348,19 +2355,27 @@ void WaveTrack::Join(
 {
    assert(IsLeader());
-   const auto channels = TrackList::Channels(this);
-   // Merge all WaveClips overlapping selection into one
```
That comment line might be kept

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2348,19 +2355,27 @@ void WaveTrack::Join(
       reportProgress);
+
+   for (const auto pChannel : TrackList::Channels(this))
+      JoinOne(*pChannel, t0, t1);
```
I see this commit that changes join behavior, and I ask whether it might make unnecessary changes in case clips already have equal stretches.

saintmatthieu replied:
Ah, crap, I forgot that check ... Thank you !

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -659,23 +659,17 @@ void WaveClip::WriteXML(XMLWriter &xmlFile) const
    if (GetWidth() != other.GetWidth())
       return false;
 
+   assert(GetStretchRatio() == other.GetStretchRatio());
```
If the function can already return false for failure, and uses of it are checked, maybe there should not be an assertion.

saintmatthieu replied:
Yeah sure.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2193,7 +2197,12 @@ void WaveTrack::Paste(double t0, const Track &src)
+      // Currently `Paste` isn't used by code that wants the newer "don't merge
+      // when copy/pasting" behaviour ...
+      constexpr auto merge = false;
+      PasteWaveTrack(t0, *other, merge);
```
I'm confused.  `true` not `false` must be passed if the intention is NOT to change behavior in this path, as the comment suggests.  That is, to pass the default for `split` to `PasteOne`, which is `true`, but is also the negation of `merge`.

Places where you can reach this: 
- Repeat effect
- Paulstretch
- Cut, copy, split cut by label
- Track::SyncLockAdjust (but is this in fact unreachable because of overriding functions?)

Other places, but where I think the destination track is always empty so it does not matter:
- Stereo to mono
- Benchmark

saintmatthieu replied:
Right, should be `true`.
Your mention of Paulstretch made me uncover another bug, where when applied to a stretched clip results in the not-enough-room error. I'll have a fix commit for that.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1452,7 +1452,7 @@ void WaveTrack::ClearAndPasteOne(
 
    // And paste in the new data
-   PasteOne(track, t0, src, startTime, endTime, split);
+   PasteOne(track, t0, src, startTime, endTime, !split);
```
Reviewing this commit, I looked at all uses of `PasteOne`, and there is one call that was not rewritten, because it passes a defaulted value.

That is in sync-lock adjustment of a WaveTrack to lengthen it with silence.

Was this case tested?  Is the desired behavior to insert silence without splits, or with them?  Is this consistent with other insertions of silence, like the Silence Audio command?

Paul-Licameli replied:
The commit that made this change fails to compile because one more use of `split` remains in `WaveTrack::PasteOne`

saintmatthieu replied:
> Reviewing this commit, I looked at all uses of PasteOne, and there is one call that was not rewritten, because it passes a defaulted value.

The default value was negated, and so was its usage in `PasteOne`, so all good here. The intention was from the very start not to change the behaviour for existing usages, so I haven't reconsidered them.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -1842,7 +1842,8 @@ void WaveTrack::HandleClear(
    if (!moveClipsLeft && shiftClipAtT1ToT0)
-      if (const auto clip = GetClipAtTime(t1))
+      if (const auto clip = GetClipAtTime(t1);
+          clip && clip->GetPlayStartTime() == t1)
```
I see an exact comparison of floating point values, and I want to know whether they were first rounded to a multiple of the same inverted rate.  That's not obvious in context here.

Paul-Licameli replied:
And:  please amend the comment for this commit, with some line breaks

saintmatthieu replied:
> I see an exact comparison of floating point values, and I want to know whether they were first rounded to a multiple of the same inverted rate. That's not obvious in context here.

You're right, adding some rounding at the beginning of that method.

saintmatthieu replied:
> And: please amend the comment for this commit, with some line breaks

Where do you see the need for line breaks?

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2000,14 +2000,16 @@ void WaveTrack::PasteOne(
+    else
     {
-       track.SplitAt(t0);
+       if (!merge)
```
This repairs the build failure of two commits previous

Paul-Licameli replied:
So, behavior may change in the case editing can't move clips, and merge is true.  The lines after this small-scope `if` happen when they did not happen before.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2000,14 +2000,16 @@ void WaveTrack::PasteOne(
        if (!track.IsEmpty(t, t + insertDuration))
           throw notEnoughSpaceException;
-       if (clipAtT0)
+       if (clipAtT0 && clipAtT0->GetPlayStartTime() == t0)
```
Same question here as in the previous commit:  should times be rounded before comparison?

saintmatthieu replied:
They probably should. Addressing ...

Paul-Licameli commented:
> Thanks @dozzzzer. I noticed yet other problems: at the end of a clip,
> 
> ![image](https://user-images.githubusercontent.com/22740106/272072568-3f6212cb-8383-4010-9c8d-38384ddd3a6b.png)
> 
> generate a chirp. You get this ugly result:
> 
> ![image](https://user-images.githubusercontent.com/22740106/272072621-722a5e14-39e2-44ae-8734-2b589cdac7ba.png)
> 
> Unfortunately I didn't manage solving this before my day off tomorrow. @Paul-Licameli if you have the capacity to look into this. Clue: the `PasteTimeWarper` is causing problems when restoring the splits in `ClearAndPasteOne`.

I attempted no debugging of this.  Have you fixed it?

Paul-Licameli commented on src/effects/nyquist/Nyquist.cpp:
```diff
@@ -51,35 +51,36 @@ effects from this one class.
-#include "../../tracks/playabletrack/wavetrack/ui/WaveChannelView.h"
-#include "../../tracks/playabletrack/wavetrack/ui/WaveChannelViewConstants.h"
-#include "../../widgets/NumericTextCtrl.h"
-#include "ProgressDialog.h"
```
Remember, I prefer to keep include directives sorted by the last path component only, because path prefixes will often change (shorten) as we make more libraries.  So omit this please.

Paul-Licameli reviewed (changes requested):
Regression in exception safety should be fixed

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -1281,6 +1280,8 @@ void WaveClip::ApplyStretchRatio(
    SetTrimRight(0.);
    SetSequenceStartTime(originalPlayStartTime);
+   mRawAudioTempo = *mProjectTempo;
+   mClipStretchRatio = targetRatio;
```
This change is wrong for exception safety!

Please keep the finally happy-path updates as they were, inside the `Finally`.

`Flush` can throw, and we can't predict what overrides of `Invalidate` may throw.

Paul-Licameli replied:
Also please preserve the assertion of the postcondition.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -1194,20 +1196,17 @@ void WaveClip::ApplyStretchRatio(
-      {
-         this->mClipStretchRatio = targetRatio;
-         this->mRawAudioTempo = this->mProjectTempo;
-         assert(GetStretchRatio() == targetRatio);
```
Ignore what I said about exception safety:  this is still right.

But I would like to conserve this assertion of the postcondition.

Paul-Licameli commented on libraries/lib-effects/PerTrackEffect.cpp:
```diff
@@ -320,7 +320,8 @@ bool PerTrackEffect::ProcessPass(TrackList &outputs,
             const auto t1 = ViewInfo::Get(*FindProject()).selectedRegion.t1();
-            PasteTimeWarper warper{ t1, mT0 + wt.GetEndTime() };
+            PasteTimeWarper warper { t1,
+                                     mT0 + (*results->begin())->GetEndTime() };
```
Yes, this corrects a mistake I introduces at af5911aa2e59f1ebcdc0c4e5e8c8e9581fb49dd5

Paul-Licameli commented:
My review of new commits is complete

saintmatthieu commented:
> > Thanks @dozzzzer. I noticed yet other problems: at the end of a clip,
> > ![image](https://user-images.githubusercontent.com/22740106/272072568-3f6212cb-8383-4010-9c8d-38384ddd3a6b.png)
> > generate a chirp. You get this ugly result:
> > ![image](https://user-images.githubusercontent.com/22740106/272072621-722a5e14-39e2-44ae-8734-2b589cdac7ba.png)
> > Unfortunately I didn't manage solving this before my day off tomorrow. @Paul-Licameli if you have the capacity to look into this. Clue: the `PasteTimeWarper` is causing problems when restoring the splits in `ClearAndPasteOne`.
> 
> I attempted no debugging of this. Have you fixed it?

Yes.

saintmatthieu commented:
@Paul-Licameli finished addressing your comments.

saintmatthieu commented:
Now looking into the [strange, suboptimal behaviour reported by @petersampsonaudacity](ttps://github.com/audacity/audacity/issues/5273#issuecomment-1748816257)

dozzzzer commented:
Tested. No new issues were found.

Paul-Licameli commented on src/effects/Paulstretch.cpp:
```diff
@@ -167,8 +168,10 @@ bool EffectPaulstretch::Process(EffectInstance &, EffectSettings &)
-         track->Paste(t0, *tempList);
+         PasteTimeWarper warper { t1, t0 + (*tempList->begin())->GetEndTime() };
+         constexpr auto preserve = false;
+         constexpr auto merge = true;
```
Why is `preserve` non-default?  Other effects using a time warper pass the default `true` -- except Nyquist, which lets the Lisp code choose.

And now I also see that NoiseReduction uses non-default for `merge` and I question that other inconsistency.

saintmatthieu replied:
Just being extra cautious: the code before didn't preserve splits, and I didn't want to change that behaviour here.

saintmatthieu replied:
Motivation in NoiseReduction.cpp would have been the same, I should think.

Paul-Licameli commented on libraries/lib-wave-track/WaveClip.cpp:
```diff
@@ -664,7 +664,14 @@ bool WaveClip::Paste(double t0, const WaveClip& other)
+   {
+      // Empty clip: we're flexible and adopt the other's stretching.
+      mRawAudioTempo = other.mRawAudioTempo;
+      mClipStretchRatio = other.mClipStretchRatio;
```
Must you also copy `mProjectTempo` to be sure `GetStrectchRatio` returns the same for both clips?

saintmatthieu replied:
I don't expect there to be a use case for that: an empty clip normally is something freshly created for future insertion into a track, at which stage it gets the project tempo update.
But it wouldn't hurt to be extra cautious. Will add a to-be-squashed commit for that.

Paul-Licameli commented on libraries/lib-wave-track/WaveTrack.cpp:
```diff
@@ -2385,12 +2386,21 @@ void WaveTrack::Join(
+             return first != interval->GetStretchRatio();
+          }))
+      BasicUI::SplitProgress(
+         intervalsToJoin.begin(), intervalsToJoin.end(),
```
This part looks good -- but you might even add `+ 1` to this `begin`

Paul-Licameli replied:
Oh I'm a fool

Paul-Licameli commented:
Fini!

saintmatthieu commented:
Likewise :)

saintmatthieu commented:
> > Likewise :)
> 
> Will I see this soon?
> 
> > I don't expect there to be a use case for that: an empty clip normally is something freshly created for future insertion into a track, at which stage it gets the project tempo update. But it wouldn't hurt to be extra cautious. Will add a to-be-squashed commit for that.

![image](https://github.com/audacity/audacity/assets/22740106/a4fa931e-6ce6-406b-9314-1c8d0bdaae90)

Wanted to be eco-friendly saving a build by waiting for rebase.
